README.md
# MVM
## MVM Architecture
List of available instructions:
| Op. code | Op. mnemo  | Description                                                                    |
|:--------:|:-----------|:-------------------------------------------------------------------------------|
| ``0x00`` | ``NOP``    | No operation                                                                   |
| ``0x01`` | ``HALT``   | Stop the program execution                                                     |
| ``0x02`` | ``PUSH x`` | Push number to stack                                                           |
| ``0x03`` | ``POP``    | Pop number from a stack                                                        |
| ``0x04`` | ``DUP``    | Duplicate the top stack element                                                |
| ``0x05`` | ``SWAP``   | Swap first and second stack elements                                           |
| ``0x06`` | ``ADD``    | Pop 1st and 2nd stack elements, add them and push to stack                     |
| ``0x07`` | ``SUB``    | Pop 1st and 2nd stack elements, subtract them and push to stack                |
| ``0x08`` | ``MUL``    | Pop 1st and 2nd stack elements, multiply them and push to stack                |
| ``0x09`` | ``DIV``    | Pop 1st and 2nd stack elements, divide them and push to stack                  |
| ``0x0A`` | ``NEG``    | Negate the top stack element                                                   |
| ``0x0B`` | ``NOT``    | Pop top stack element. If it's equal to ``0`` then push ``1``, else push ``0`` |
| ``0x0C`` | ``CALL x`` | Jump to the ``x`` address and push new stack frame                             |
| ``0x0D`` | ``RET``    | Return to the caller and pop the stack frame                                   |
| ``0x0E`` | ``JMP x``  | Perform the unconditional jump to ``x`` address                                |
| ``0x0F`` | ``JE x``   | Jump to ``x`` **if** top element ``== 0``                                      |
| ``0x10`` | ``JNE x``  | Jump to ``x`` **if** top element ``!= 0``                                      |
| ``0x11`` | ``JG x``   | Jump to ``x`` **if** top element ``> 0 ``                                      |
| ``0x12`` | ``JL x``   | Jump to ``x`` **if** top element ``< 0 ``                                      |
| ``0x13`` | ``JGE x``  | Jump to ``x`` **if** top element ``>= 0``                                      |
| ``0x14`` | ``JLE x``  | Jump to ``x`` **if** top element ``<= 0``                                      |
| ``0x15`` | ``LD x``   | Push local variable to stack                                                   |

CHANGELOG.md
# Revision history for MVM

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


app/Instruction.hs
module Instruction (
  Op(..),
  Instruction(..),
  Command(..),
  instructions,
  instructionByOp,
  toOp
) where

import Data.Foldable
import qualified Data.Char as Char
import qualified Data.Map as Map
import qualified Data.Sequence as S

import qualified VirtualMachine as VM

data Op = Nop  -- 0x00 
        | Halt -- 0x01
        | Push -- 0x02
        | Pop  -- 0x03
        | Dup  -- 0x04
        | Swap -- 0x05
        | Add  -- 0x06
        | Sub  -- 0x07
        | Mul  -- 0x08
        | Div  -- 0x09
        | Neg  -- 0x0a
        | Not  -- 0x0b
        | Call -- 0x0c
        | Ret  -- 0x0d
        | Jmp  -- 0x0e
        | Je   -- 0x0f
        | Jne  -- 0x10
        | Jg   -- 0x11
        | Jl   -- 0x12
        | Jge  -- 0x13
        | Jle  -- 0x14
        | Ld   -- 0x15
        deriving (Eq, Ord, Enum, Show, Read, Bounded)

type Args = [Int]
type Pops = [Int]
type Pushes = S.Seq Int

data Instruction = Simple     { op        :: Op
                              , noParams  :: Int
                              , noPops    :: Int
                              , sAction   :: Args -> Pops -> Pushes
                              } 
                   | Complex  { op        :: Op
                              , noParams  :: Int
                              , cAction   :: VM.VM -> Args -> Either String VM.VM
                              } 

data Command = Command { instr  :: Instruction
                       , args   :: [Int]
                       } 

instructions :: [Instruction]
instructions = [ Simple  { op = Nop,  noParams = 0, noPops = 0, sAction = (\_ _      -> S.empty)                               }
               , Simple  { op = Push, noParams = 1, noPops = 0, sAction = (\args _   -> S.fromList args)                       }
               , Simple  { op = Pop,  noParams = 0, noPops = 1, sAction = (\_ _      -> S.empty)                               }
               , Simple  { op = Dup,  noParams = 0, noPops = 1, sAction = (\_ [x]    -> S.fromList [x, x])                     }
               , Simple  { op = Swap, noParams = 0, noPops = 2, sAction = (\_ [x, y] -> S.fromList [y, x])                     }
               , Simple  { op = Add,  noParams = 0, noPops = 2, sAction = (\_ [x, y] -> S.fromList [y + x])                    }
               , Simple  { op = Sub,  noParams = 0, noPops = 2, sAction = (\_ [x, y] -> S.fromList [y - x])                    }
               , Simple  { op = Mul,  noParams = 0, noPops = 2, sAction = (\_ [x, y] -> S.fromList [y * x])                    }
               , Simple  { op = Div,  noParams = 0, noPops = 2, sAction = (\_ [x, y] -> S.fromList [y `div` x])                }
               , Simple  { op = Neg,  noParams = 0, noPops = 1, sAction = (\_ [x]    -> S.fromList [-x])                       }
               , Simple  { op = Not,  noParams = 0, noPops = 1, sAction = (\_ [x]    -> S.fromList [if x /= 0 then 0 else 1])  }
               , Complex { op = Halt, noParams = 0, cAction = (\vm _   -> Right $ vm { VM.halt = True })                       }
               , Complex { op = Jmp,  noParams = 1, cAction = (\vm [x] -> Right $ vm { VM.pc = x})                             }
               , Complex { op = Je,   noParams = 1, cAction = jumpIf (==)                                                      }
               , Complex { op = Jne,  noParams = 1, cAction = jumpIf (/=)                                                      }
               , Complex { op = Jg,   noParams = 1, cAction = jumpIf (>)                                                       }
               , Complex { op = Jl,   noParams = 1, cAction = jumpIf (<)                                                       }
               , Complex { op = Jge,  noParams = 1, cAction = jumpIf (>=)                                                      }
               , Complex { op = Jle,  noParams = 1, cAction = jumpIf (<=)                                                      }
               ]

jumpIf :: (Int -> Int -> Bool) -> VM.VM -> Args -> Either String VM.VM
jumpIf predicate vm [addr] = Right $ vm { VM.pc = pc' }
  where
    (top:_) = toList . VM.stack $ vm
    pc' = if top `predicate` 0 then addr else VM.pc vm + 1

instructionByOp :: Map.Map Op Instruction
instructionByOp = Map.fromList $ map (\i -> (op i, i)) instructions

toOp :: String -> Op
toOp = read . capitalize
  where capitalize :: String -> String
        capitalize [] = []
        capitalize (x:xs) = Char.toUpper x : map Char.toLower xs